src/net/test.rs

```diff
@@ -11,17 +22,134 @@ use axum::{
     http::Request,
 };
 use futures::Stream;
-use hyper::{
-    client::HttpConnector,
-    http::{uri::Scheme, Uri},
-};
-use hyper_tls::{native_tls::TlsConnector, HttpsConnector};
+use hyper_tls::native_tls::Identity;
 use once_cell::sync::Lazy;
-use std::{array, error::Error as StdError, net::SocketAddr, ops::Deref};
+use std::{
+    array,
+    error::Error as StdError,
+    net::{SocketAddr, TcpListener},
+};
 use tokio::task::JoinHandle;
+use tokio_rustls::rustls::Certificate;
+
+pub const DEFAULT_TEST_PORTS: [u16; 3] = [3000, 3001, 3002];
 
-static DEFAULT_CLIENT_CONFIG: Lazy<PeerConfig> =
-    Lazy::new(|| PeerConfig::new("http://localhost:3000".parse().unwrap()));
+pub struct TestConfig {
+    pub disable_https: bool,
+    pub network: NetworkConfig,
+    pub servers: [ServerConfig; 3],
+    pub sockets: Option<[TcpListener; 3]>,
+}
+
+impl TestConfig {
+    #[must_use]
+    pub fn builder() -> TestConfigBuilder {
+        TestConfigBuilder::default()
+    }
+}
+
+impl Default for TestConfig {
+    fn default() -> Self {
+        Self::builder().build()
+    }
+}
+
+#[must_use]
+fn server_config_insecure_http(port: u16) -> ServerConfig {
+    ServerConfig {
+        port: Some(port),
+        disable_https: true,
+        tls: None,
+    }
+}
+
+#[must_use]
+pub fn server_config_https(id: HelperIdentity, port: u16) -> ServerConfig {
+    let (certificate, private_key) = get_test_certificate_and_key(id);
+    ServerConfig {
+        port: Some(port),
+        disable_https: false,
+        tls: Some(TlsConfig::Inline {
+            certificate: String::from_utf8(certificate.to_owned()).unwrap(),
+            private_key: String::from_utf8(private_key.to_owned()).unwrap(),
+        }),
+    }
+}
+
+#[derive(Default)]
+pub struct TestConfigBuilder {
+    ports: Option<[u16; 3]>,
+    disable_https: bool,
+}
+
+impl TestConfigBuilder {
+    #[must_use]
+    pub fn with_http_and_default_test_ports() -> Self {
+        Self {
+            ports: Some(DEFAULT_TEST_PORTS),
+            disable_https: true,
+        }
+    }
+
+    #[must_use]
+    pub fn with_open_ports() -> Self {
+        Self {
+            ports: None,
+            disable_https: false,
+        }
+    }
+
+    #[must_use]
+    pub fn with_disable_https_option(mut self, value: bool) -> Self {
+        self.disable_https = value;
+        self
+    }
+
+    #[must_use]
+    pub fn build(self) -> TestConfig {
+        let mut sockets = None;
+        let ports = self.ports.unwrap_or_else(|| {
+            let socks = array::from_fn(|_| TcpListener::bind("127.0.0.1:0").unwrap());
```

Comment:
The IPv6 people I know would be disappointed here.  But that's their problem, not ours :)

Reply:
I can try changing this (and a few other places IPv4 localhost is used) to the name `localhost`. I do generally think it's worth doing the right thing for IPv6 when the choice presents itself. I'm not sure I'd want to bet money on whether this code sees a machine without an IPv4 stack or a machine with broken name resolution of localhost first, but we can see.